### PR TITLE
Added serialization for ISOTPSockets

### PR DESCRIPTION
**Pull request by Feromrk**
This is a pretty small PR.

Added functions for serializing ISOTPSockets into dicts. Currently I am working on a project where I need to store those sockets in JSON format. Well not really the sockets, but the information that is necessary to create those sockets from the stored JSON.

### CODE REVIEW

**Comment by Feromrk** [comment]
I will take a closer look on the tests, probably tomorrow.

**Comment by polybassa** [comment]
I'm fine with this PR. What about a unit test for the recreation of a socket?

-- 
Viele Grüße

Nils Weiß

**Comment by Feromrk** [comment]
@guedou my workflow in my project is like this:
```
# do some scapy stuff, e.g. scanning
# create list of isotpsockets

# create a dict with results which is later converted to JSON by an external API

# call the (ansible) API, which takes keyword arguments and converts them to json
# this call is mandatory
module.exit_json(**result)
```
if I would create directly JSON documents from the ISOTPSockets I would need to parse JSON and in the end convert back to JSON. So this is dependent on my needs. Nothing really serious though. 

Do you think `to_JSON` would suit scapy better than `to_dict`?

File: scapy/contrib/isotp.py [hunk 1]
```diff
@@ -656,6 +657,14 @@ def __init__(self,
         self.src = sid
         self.dst = did
 
+        if hasattr(can_socket, "to_dict"):
+            self.sockd = can_socket.to_dict()
```
Comment (polybassa) [inline scapy/contrib/isotp.py]: Could you try to call `to_dict()` only when you need it

**Comment by guedou** [comment]
If this PR is only useful for your project, I doubt that it should be integrated into Scapy.

Serializing to JSON is a better option that will provide a more stable output. To my understanding, the dict output might change between Python releases.

**Comment by polybassa** [comment]
I agree, this code might not be useful for a lot of people

**Comment by polybassa** [comment]
@guedou  What do you think about a `command()` call for sockets? Similar to the call in `Packet`

**Comment by Feromrk** [comment]
@guedou I dont't think that having a serialization function is a bad idea. There are if course many possibilities, be it JSON, a python dict, a string or whatever. As @polybassa said, there seems to be a function for Packets. Is it a bad idea to add a similar feature to sockets?

**Comment by guedou** [comment]
@polybassa a `command()` method for SuperSocket is a really good solution! I like the idea.

**Comment by Feromrk** [comment]
@guedou what do you think about the changes?

**Comment by gpotter2** [comment]
Could you rebase & squash your PR?

**Comment by Feromrk** [comment]
@polybassa after a rebase those CCP tests started failing (I don't think this is related to my code). Do you have an idea whats going on there?

File: scapy/contrib/cansocket_native.py [hunk 2]
```diff
@@ -13,6 +14,7 @@
 import struct
 import socket
 import time
+import six
```
Comment (polybassa) [inline scapy/contrib/cansocket_native.py:17]: Please use `import scapy.modules.six as six`

File: scapy/contrib/cansocket_native.py [hunk 3]
```diff
@@ -74,6 +79,47 @@ def __init__(self, channel=None, receive_own_messages=False,
...
+            receive_own_messages=self.receive_own_messages,
+            can_filters=self.can_filters,
+            remove_padding=self.remove_padding,
+            basecls=self.basecls
+        )
+
+        dnv = []
+        for dn, dv in six.iteritems(d):
+            if (dv is None or
+                    (isinstance(dv, (list, dict, set)) and len(dv) == 0)):
```
Comment (polybassa) [inline scapy/contrib/cansocket_native.py:94]: you could use `hasattr(dv, "__iter__")` to achieve the same, but more general

File: scapy/contrib/cansocket_native.py [hunk 4]
```diff
@@ -74,6 +79,47 @@ def __init__(self, channel=None, receive_own_messages=False,
...
+                dv = "[%s]" % ",".join(
+                    "{%s}" % ",".join(
+                        can_filter_g(filter)
+                    ) for filter in dv
+                )
+            else:
+                dv = repr(dv)
+            dnv.append("%s=%s" % (dn, dv))
+
+        clsname = (
```
Comment (polybassa) [inline scapy/contrib/cansocket_native.py:116]: you should use `self.__class__.__name__` here. I don't see a case where the class name is in `self.__name__`

File: scapy/contrib/cansocket_native.py [hunk 5]
```diff
@@ -74,6 +79,47 @@ def __init__(self, channel=None, receive_own_messages=False,
         self.ins.bind((self.channel,))
         self.outs = self.ins
 
+    def command(self):
```
Comment (polybassa) [inline scapy/contrib/cansocket_native.py:82]: all your command functions look like very redundant code. Could you try to not repeat yourself.

File: scapy/contrib/cansocket_native.py (hunk 3 above)
Comment (Feromrk) [reply scapy/contrib/cansocket_native.py:94]: then, I need to add a try and except TypeError
```
>>> hasattr(ISOTP, "__iter__")                                                  
True
>>> hasattr(ISOTP, "__len__")                                                   
True
>>> len(ISOTP)                                                                  
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-e63da5f1917e> in <module>
----> 1 len(ISOTP)

TypeError: object of type 'Packet_metaclass' has no len()
```

Comment (polybassa) [reply scapy/contrib/cansocket_native.py:94]: Sorry, I misunderstood your use case.. I thought you wanted to test for containers

Comment (Feromrk) [reply scapy/contrib/cansocket_native.py:94]: I do, that's why I like your solution. I will implement it with the stated try, except.

Comment (polybassa) [reply scapy/contrib/cansocket_native.py:94]: or exclude if it is instance of `Packet`

Comment (polybassa) [reply scapy/contrib/cansocket_native.py:94]: or derives from metaclass for Packet

**Comment by Feromrk** [comment]
@polybassa thanks for your review. I implemented your suggestions. Please review again =)

File: scapy/contrib/cansocket_python_can.py [hunk 6]
```diff
@@ -141,9 +142,13 @@ class PythonCANSocket(SuperSocket, SelectableObject):
     nonblocking_socket = True
 
     def __init__(self, **kwargs):
+        self.args = copy.deepcopy(kwargs)
```
Comment (polybassa) [inline scapy/contrib/cansocket_python_can.py:145]: I think this line breaks the CCP unit tests. Maybe you can check the basecls parameter. This leads probably to the error. I assume the basecls isn't set correctly in the test cases anymore.

Comment (Feromrk) [reply scapy/contrib/cansocket_python_can.py:145]: Interesting, I will try another method to see if anything changes.

Comment (Feromrk) [reply scapy/contrib/cansocket_python_can.py]: After a lot of testing I found out, that the breaking part happens not here, but in isotp.uts. the python builtin exec() that I use for reconstructing the ISOTPSockets breaks stuff in the interpreter but only in python2 and pypy (it is implementation dependent). I tried to somehow get it work in py2 and pypy but couldn't find a way, so I just disabled the test there.

File: scapy/supersocket.py [hunk 7]
```diff
@@ -185,6 +185,55 @@ def tshark(self, *args, **kargs):
...
+                continue
+            elif hasattr(v, "__iter__"):
+                try:
+                    c = len(v) == 0
+                except TypeError:
+                    c = False
+                if c:
+                    continue
+
+            # provide a callback if all other cases,
```
Comment (polybassa) [inline scapy/supersocket.py:212]: If you don't use the callbacks, better remove them.

**Comment by polybassa** [comment]
besides my comment, I approve. Could you please rebase and squash.

File: scapy/supersocket.py (hunk 7 above)
Comment (Feromrk) [reply scapy/supersocket.py:212]: Right

**Comment by Feromrk** [comment]
@polybassa @gpotter2 Rebase and squash done.

**Comment by Feromrk** [comment]
@guedou Please give your opinion =)

**Comment by guedou** [comment]
Is it possible to have a generic implementation for `SuperSocket`? Also, could you change the one letter variables to something more meaningful? The one letter variable will likely break future `flake8` tests.

**Comment by Feromrk** [comment]
A generic implementation would require refactoring of all available types of sockets. Every class that inherits `SuperSocket` would need to store the parameter names and values of its constructor in a predefined manner. Since I have no clue if that is possible for every socket, I don't think that is a good idea.
I just renamed those variables.

**Comment by gpotter2** [comment]
In all honesty (I know you've put work into this, sorry for that), I only see code to maintain that will likely be never used.

I fail to see a use case for this :/ a socket is the last thing you want to serialize, and if you do you probably should do so within your own code..

**Comment by Feromrk** [comment]
@gpotter2 would have been cool, if you would have said that earlier. I've put some hours into this. But well, it is what it is.
@guedou is this your final decision?

**Comment by guedou** [comment]
If this serialization cannot be generalized to other `SuperSocket` objects, I think that the changes should not be part of Scapy. You could for example move this code to another git repository and reference it from either Scapy documentation or https://github.com/secdev/awesome-scapy

**Pull request closed without merging**